apps/prairielearn/src/lib/editors.js

```diff
@@ -56,425 +92,106 @@ class Editor {
-        courseUtil.getOrUpdateCourseCommitHash(this.course, (err, hash) => {
-          ERR(err, (e) => logger.error('Error in updateCourseCommitHash()', e));
-          startGitHash = hash;
+            const lockName = `coursedir:${this.course.path}`;
+            await namedLocks.doWithLock(lockName, { timeout: 5000 }, async () => {
+              const startGitHash = await courseUtil.getOrUpdateCourseCommitHashAsync(this.course);
 
-          if (config.fileEditorUseGit) {
-            _clean(_write, _cleanup);
-          } else {
-            _write();
-          }
-        });
-      };
-
-      const _clean = (on_success, on_error) => {
-        debug(`${job_sequence_id}: _clean`);
-        const jobOptions = {
-          course_id: options.course_id,
-          user_id: options.user_id,
-          authn_user_id: options.authn_user_id,
-          job_sequence_id: job_sequence_id,
-          type: 'clean_git_repo',
-          description: 'Clean local files not in remote git repository',
-          command: 'git',
-          arguments: ['clean', '-fdx'],
-          working_directory: this.course.path,
-          env: gitEnv,
-          on_success: () => {
-            _reset(on_success, on_error);
-          },
-          on_error: on_error,
-          no_job_sequence_update: true,
-        };
-        serverJobs.spawnJob(jobOptions);
-      };
-
-      const _reset = (on_success, on_error) => {
-        debug(`${job_sequence_id}: _reset`);
-        const jobOptions = {
-          course_id: options.course_id,
-          user_id: options.user_id,
-          authn_user_id: options.authn_user_id,
-          job_sequence_id: job_sequence_id,
-          type: 'reset_from_git',
-          description: 'Reset state to remote git repository',
-          command: 'git',
-          arguments: ['reset', '--hard', `origin/${this.course.branch}`],
-          working_directory: this.course.path,
-          env: gitEnv,
-          on_success: on_success,
-          on_error: on_error,
-          no_job_sequence_update: true,
-        };
-        serverJobs.spawnJob(jobOptions);
-      };
-
-      const _write = () => {
-        debug(`${job_sequence_id}: _write`);
-        const jobOptions = {
-          course_id: options.course_id,
-          user_id: options.user_id,
-          authn_user_id: options.authn_user_id,
-          type: 'write',
-          description: 'Write to disk',
-          job_sequence_id: job_sequence_id,
-          on_success: config.fileEditorUseGit ? _add : _syncFromDisk,
-          on_error: config.fileEditorUseGit ? _cleanupAfterWrite : _cleanup,
-          no_job_sequence_update: true,
-        };
-        serverJobs.createJob(jobOptions, (err, job) => {
-          if (ERR(err, (e) => logger.error('Error in createJob()', e))) {
-            _finishWithFailure();
-            return;
-          }
+              if (config.fileEditorUseGit) {
+                await cleanAndResetRepository(this.course, gitEnv, job);
```

Comment:
`instructorFileEditor.js` will do a `git fetch` before this - is it worth adding the same here?

Reply:
On a second look, it only actually fetches if the user explicitly requests that after a failed push; grep for `pull_and_save_and_sync`. It would be great to have consistent behavior here, but I think that may be outside the scope of this PR.